File: arena-ffi/src/dependency/oracle/oracle_dependency.rs
use arena::Dependency;
use arena_oracledb::OracleDependency;
use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct OracleDependencyConfig {
    pub identifier: String,
    #[serde(default)]
    pub image_name: Option<String>,
    #[serde(default)]
    pub image: Option<String>,
    #[serde(default)]
    pub port: Option<u16>,
    #[serde(default)]
    pub database_name: Option<String>,
    #[serde(default)]
    pub database_username: Option<String>,
    #[serde(default)]
    pub database_password: Option<String>,
    #[serde(default)]
    pub admin_password: Option<String>,
    #[serde(default)]
    pub container_name: Option<String>,
    #[serde(default)]
    pub startup_sql_scripts: Option<Vec<String>>,
}

pub fn build(config: &OracleDependencyConfig, network: Option<&str>) -> Result<Dependency, String> {
    let mut builder = OracleDependency::builder(&config.identifier);
    if let Some(image) = config.image.as_deref() {
        builder = builder.with_image(image);
    }
    if let Some(ref image_name) = config.image_name {
        builder = builder.with_image_name(image_name);
    }
    if let Some(port) = config.port {
        builder = builder.with_port(port);
    }
    if let Some(ref database_name) = config.database_name {
        builder = builder.with_database_name(database_name);
    }
    if let Some(ref database_username) = config.database_username {
        builder = builder.with_database_username(database_username);
    }
    if let Some(ref database_password) = config.database_password {
        builder = builder.with_database_password(database_password);
    }
    if let Some(ref admin_password) = config.admin_password {
        builder = builder.with_admin_password(admin_password);
    }
    if let Some(ref scripts) = config.startup_sql_scripts {
        builder = builder.with_startup_sql_scripts(scripts.clone());
    }
    if let Some(network) = network {
        builder = builder.with_network(network);
    }
    if let Some(ref container_name) = config.container_name {
        builder = builder.with_container_name(container_name);
    }
    Ok(Box::new(builder.build()))
}

File: arena-xunit/src/Dep/OracleDependency.cs
using System;
using System.Collections.Generic;
using ArenaDotnet.Xunit.Support;
using Newtonsoft.Json.Linq;

namespace ArenaDotnet.Xunit.Dep;

public sealed class OracleDependency : IArenaDependency
{
    public string Type => "oracledb";
    public string Identifier { get; }
    public long? ExpirySeconds { get; internal set; }
    public int Port { get; }
    public string? DatabaseName { get; }
    public string? DatabaseUsername { get; }
    public string? DatabasePassword { get; }
    public string? AdminPassword { get; }
    public string? ImageName { get; }
    public string? Image { get; }
    public string? ContainerName { get; }
    public IReadOnlyList<string> StartupSqlScripts { get; }
    public string? SetupMode { get; }
    public long? SqlReadinessTimeoutMs { get; }
    public List<JToken>? Children => ChildrenWireFormat.Build(_children);

    private readonly IReadOnlyList<IArenaDependency> _children;

    internal OracleDependency(
        string identifier,
        int port,
        string? databaseName,
        string? databaseUsername,
        string? databasePassword,
        string? adminPassword,
        string? imageName,
        string? image,
        string? containerName,
        IReadOnlyList<string> startupSqlScripts,
        string? setupMode,
        long? sqlReadinessTimeoutMs,
        IReadOnlyList<IArenaDependency> children)
    {
        Identifier = identifier;
        Port = port;
        DatabaseName = databaseName;
        DatabaseUsername = databaseUsername;
        DatabasePassword = databasePassword;
        AdminPassword = adminPassword;
        ImageName = imageName;
        Image = image;
        ContainerName = containerName;
        StartupSqlScripts = startupSqlScripts;
        SetupMode = setupMode;
        SqlReadinessTimeoutMs = sqlReadinessTimeoutMs;
        _children = children;
    }

    public string ForFfi()
    {
        return ArenaJson.Serialize(this);
    }
}

public sealed class OracleDependencyBuilder
{
    private long? _expirySeconds;
    private readonly string _name;
    private int _port = 1521;
    private string? _databaseName;
    private string? _databaseUsername;
    private string? _databasePassword;
    private string? _adminPassword;
    private string? _imageName;
    private string? _image;
    private string? _containerName;
    private readonly List<string> _startupSqlScripts = new();
    private string? _setupMode;
    private long? _sqlReadinessTimeoutMs;
    private readonly List<IArenaDependency> _children = new();

    public OracleDependencyBuilder(string name)
    {
        _name = name;
    }

    public OracleDependencyBuilder WithPort(int port)
    {
        _port = port;
        return this;
    }

    public OracleDependencyBuilder WithDatabaseName(string databaseName)
    {
        _databaseName = databaseName;
        return this;
    }

    public OracleDependencyBuilder WithDatabaseUsername(string databaseUsername)
    {
        _databaseUsername = databaseUsername;
        return this;
    }

    public OracleDependencyBuilder WithDatabasePassword(string databasePassword)
    {
        _databasePassword = databasePassword;
        return this;
    }

    public OracleDependencyBuilder WithAdminPassword(string adminPassword)
    {
        _adminPassword = adminPassword;
        return this;
    }

    public OracleDependencyBuilder WithImageName(string imageName)
    {
        _imageName = imageName;
        return this;
    }

    public OracleDependencyBuilder WithImage(string image)
    {
        _image = image;
        return this;
    }

    public OracleDependencyBuilder WithContainerName(string containerName)
    {
        _containerName = containerName;
        return this;
    }

    public OracleDependencyBuilder WithStartupSqlScripts(IEnumerable<string> scripts)
    {
        _startupSqlScripts.AddRange(scripts);
        return this;
    }

    public OracleDependencyBuilder FullBuild()
    {
        _setupMode = "full_build";
        return this;
    }

    public OracleDependencyBuilder WithSqlReadinessTimeout(TimeSpan timeout)
    {
        _sqlReadinessTimeoutMs = (long)timeout.TotalMilliseconds;
        return this;
    }

    public OracleDependencyBuilder AddChildDependency(IArenaDependency child)
    {
        _children.Add(child);
        return this;
    }

    public OracleDependencyBuilder WithExpiry(System.TimeSpan expiry)
    {
        _expirySeconds = ExpirySeconds(expiry);
        return this;
    }

    public OracleDependencyBuilder WithoutExpiry()
    {
        _expirySeconds = 0;
        return this;
    }

    public OracleDependency Build()
    {
        var identifier = ArenaIdentifiers.Build("arena-oracle", _name);
        var built = new OracleDependency(identifier, _port, _databaseName, _databaseUsername, _databasePassword, _adminPassword, _imageName, _image, _containerName, _startupSqlScripts, _setupMode, _sqlReadinessTimeoutMs, _children);
        built.ExpirySeconds = _expirySeconds;
        return built;
    }

    private static long ExpirySeconds(System.TimeSpan expiry)
    {
        if (expiry < System.TimeSpan.Zero)
            throw new System.ArgumentOutOfRangeException(nameof(expiry), "expiry must not be negative");
        var seconds = (long)expiry.TotalSeconds;
        return seconds == 0 && expiry > System.TimeSpan.Zero ? 1 : seconds;
    }

}
